Implement #3431 - add media quality change events

caridley opened the pull request:
## Description
Fixes #3431 

When measuring quality of experience we like to know when media quality metrics like width, height, framerate and bitrate change at the playhead.

We have implemented this by capturing media quality information in the InitSegmentReference during manifest parsing, and added a QualityObserver that captures the playback position of quality changes when init segments are appended to the source buffer and emits MediaQualityChanged events when the play head advances over a quality change point.

This is implemented only for DASH at this point.

## Screenshots (optional)
NA

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## QUESTIONS FOR SHAKA TEAM
Do we need to support HLS? If yes are the relevant metrics available when creating InitSegmentReferences?

Do we need to support this feature for offline storage and playback?

I think we should add a config field to enable/disable this feature, since it does use CPU cycles observing the playhead at relatively high frequency. Is it OK to add that?

I would like to include the current audio and video MediaQualityInfo values in the getStats() response. Is it OK to add that?

## Checklist:

- [x] I have signed the Google CLA <https://cla.developers.google.com>
- [x] My code follows the style guidelines of this project
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [  ] I have verified my change on multiple browsers on different platforms
- [x] I have run `./build/all.py` and the build passes
- [ x] I have run `./build/test.py` and all tests pass

theodab commented on lib/media/quality_observer.js:
```diff
@@ -0,0 +1,276 @@
+      const mediaQualityAtPosition =
+        shaka.media.QualityObserver.getMediaQualityAtPosition_(
+            positionInSeconds, contentMediaQuality);
+      if (mediaQualityAtPosition &&
```
You have this check here because `mediaQualitiesAreTheSame_` considers `null` to not be the same as `null`, right? I think that would be better to move that logic into `mediaQualitiesAreTheSame_`.

caridley replied:
I am changing `mediaQualitiesAreTheSame_` such that `null` is the same as `null` because `null` != `null` is confusing. I think we need to keep this check.

theodab commented on lib/media/quality_observer.js:
```diff
@@ -0,0 +1,276 @@
+ *
+ * @typedef {function(shaka.extern.MediaQualityInfo)}
+ */
+shaka.media.QualityObserver.EventListener;
```
It might be a good idea to also pass the position into this function.

caridley replied:
done

theodab commented on lib/media/segment_reference.js:
```diff
@@ -26,8 +26,10 @@ shaka.media.InitSegmentReference = class {
    * @param {?number} endByte The offset from the start of the resource to the
    *   end of the segment, inclusive.  A value of null indicates that the
    *   segment extends to the end of the resource.
+   * @param {?shaka.extern.MediaQualityInfo} mediaQuality Information about the
```
For optional members, the normal syntax would be:
```
* @param {shaka.extern.MediaQualityInfo=}
```

caridley replied:
done

theodab commented on lib/player.js:
```diff
@@ -173,6 +174,18 @@ goog.requireType('shaka.routing.Payload');
+/**
+ * @event shaka.Player.MediaQualityChangedEvent
+ * @description Fired when the media quality changes at the playhead.
+ * That may be caused by an adapatation change or a DASH period transition.
```
Typo: "adapatation" -> "adaptation"

caridley replied:
done

theodab commented on lib/player.js:
```diff
@@ -173,6 +174,18 @@ goog.requireType('shaka.routing.Payload');
+ * @event shaka.Player.MediaQualityChangedEvent
+ * @description Fired when the media quality changes at the playhead.
+ * That may be caused by an adapatation change or a DASH period transition.
+ * Separate events are emitted for audio and video contentTypes.
```
Probably worth mentioning that (at least for the moment) this only works for DASH.

caridley replied:
done

theodab reviewed (changes requested):
Overall, I think this looks good! I do have some change requests though, mostly to do with comments and naming.

theodab reviewed (changes requested):
Also, now that I think about it, it would probably be wise to add a configuration value that determines whether or not a QualityObserver will be added to the player. The QualityObserver logic could be a little heavy-duty for a presentation with a lot of periods, so it'd be nice if people who don't care about this information can turn it off.
Call it `streaming.observeQualityChanges`, perhaps?

caridley commented:
@theodab  I am planning to make these name changes to address naming confusion

MediaQualityInfo no change
QualityChange -> QualityAtPosition
ContentMediaQuality -> QualityTimeline
contentMediaQualities_ -> contentTypeQualityTimelines_
getContentMediaQuality_(contentType) -> getQualityTimeline_(contentType)
addMediaQualityChange(mediaQuality, position) ->  addQualityAtPosition(mediaQuality, position)
getMediaQualityAtPosition_(position, contentMediaQuality) -> getQualityAtPosition_(position, qualityTimeline)
purgeQualityChangesForContentType_(contentType) -> purgeQualityTimeline_(contentType)

caridley commented:
> Also, now that I think about it, it would probably be wise to add a configuration value that determines whether or not a QualityObserver will be added to the player. The QualityObserver logic could be a little heavy-duty for a presentation with a lot of periods, so it'd be nice if people who don't care about this information can turn it off. Call it `streaming.observeQualityChanges`, perhaps?

Will do

caridley commented on externs/shaka/player.js:
```diff
@@ -468,6 +468,47 @@ shaka.extern.ID3Metadata;
+ * Contains information about the quality of an audio or video media stream.
+ *
+ * @property {?number} audioSamplingRate
+ *   The audio sampling rate.
```
done

caridley commented on externs/shaka/player.js:
```diff
@@ -468,6 +468,47 @@ shaka.extern.ID3Metadata;
+ * @property {number} bandwidth
+ *   The bandwidth in bits per second.
+ * @property {string} codecs
+ *   The codecs value
```
done

caridley commented on externs/shaka/player.js:
```diff
@@ -468,6 +468,47 @@ shaka.extern.ID3Metadata;
+ *   The video height in pixels.
+ * @property {string} mimeType
+ *   The MIME type.
+ * @property {?number} numChannels
```
done

caridley commented on lib/media/quality_observer.js:
```diff
@@ -0,0 +1,276 @@
+   * @return {?shaka.extern.MediaQualityInfo}
+   * @private
+   */
+  static getMediaQualityAtPosition_(position, contentMediaQuality) {
```
Hopefully there is less confusion with the new names

caridley commented on lib/media/quality_observer.js:
```diff
@@ -0,0 +1,276 @@
+    /**
+     * @private {!Map.<string,!shaka.media.QualityObserver.ContentMediaQuality>}
+     */
+    this.contentMediaQualities_ = new Map();
```
Names have been changed to be less ambiguous

caridley commented on lib/player.js:
```diff
@@ -5586,6 +5621,38 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
+    };
+
+    const data = {
+      'mediaQuality': clone,
```
I found that the `mediaQuality` property name was being minified to `re` in the build process unless wrapped in quotation marks. Any idea why that happens?

theodab replied:
Yeah, that's a known issue. If you don't specify the dictionary values as strings, Closure Compiler minifies them. It was reported a few weeks ago as #3710.
Actually, I made a CL pretty recently that enforces conformance on this front by changing the expected type of this data object from `Object` to `Map.<string, Object>`. See https://github.com/google/shaka-player/commit/dd3748d59970de46255cd6d8b4204ee501600655. You're probably going to have to rebase onto that, now that I think about it.

joeyparrish replied:
Any object that isn't declared as part of an "extern" interface, the compiler will rename its fields.  It's sometimes not worth creating definitions for everything, so quoted property names are the lightweight option.

theodab reviewed (changes requested):
Okay, yeah, I checked the merge action and there's definitely a merge conflict in Player having to do with the event constructor. Fix that and this should be good to go, thanks.

caridley commented:
After re-basing on master many tests, including integration tests for this feature, are failing on Firefox and Safari with this error 

TypeError: window.crypto.randomUUID is not a function. (In 'window.crypto.randomUUID()', 'window.crypto.randomUUID' is undefined) in lib/util/cmcd_manager.js (line 4587)

Should I attempt to fix this or wait for a fix from someone else?

caridley commented:
> After re-basing on master many tests, including integration tests for this feature, are failing on Firefox and Safari with this error
> 
> TypeError: window.crypto.randomUUID is not a function. (In 'window.crypto.randomUUID()', 'window.crypto.randomUUID' is undefined) in lib/util/cmcd_manager.js (line 4587)
> 
> Should I attempt to fix this or wait for a fix from someone else?

Fixed by adding this to shaka-player.uncompiled.js 
 goog.require('shaka.polyfill.RandomUUID');

caridley commented:
@theodab I think I have addressed all of the code review comments. Please let me know if there is anything else I can to help get this across the finish line.

theodab commented:
Thanks for your contributions!
